### ISSUE TITLE
MPI_Type_size() is called with MPI_DATATYPE_NULL and breaks some applications

### ISSUE THREAD

**Issue by bebosudo** [darshan-hpc/darshan#698]
Hi, a user at NERSC reported that darshan is calling the MPI_Type_size() function with MPI_DATATYPE_NULL; they protect all calls to MPI_Type_size in their application with
```
if (mpi_datatype == MPI_DATATYPE_NULL) {
  size = 0;
} else {
  size = MPI_Type_size(mpi_datatype);
}
```

And he suggests to do the same in Darshan.

The following is an example traceback, but he couldn't provide more details, so I don't have a piece of code to reproduce the issue:

```
3: Test command: /usr/bin/srun "-n" "5" "/path/to/test/app"
3: Test timeout computed to be: 600
3: Rank 1 [Tue Mar 22 10:03:09 2022] [c1-0c0s6n1] Fatal error in PMPI_Type_size: Invalid datatype, error stack:
3: PMPI_Type_size(119): MPI_Type_size(MPI_DATATYPE_NULL) failed
3: PMPI_Type_size(69).: Datatype for argument datatype is a null datatype
3: Rank 2 [Tue Mar 22 10:03:09 2022] [c1-0c0s6n1] Fatal error in PMPI_Type_size: Invalid datatype, error stack:
3: PMPI_Type_size(119): MPI_Type_size(MPI_DATATYPE_NULL) failed
3: PMPI_Type_size(69).: Datatype for argument datatype is a null datatype
3: Rank 3 [Tue Mar 22 10:03:09 2022] [c1-0c0s6n1] Fatal error in PMPI_Type_size: Invalid datatype, error stack:
3: PMPI_Type_size(119): MPI_Type_size(MPI_DATATYPE_NULL) failed
3: PMPI_Type_size(69).: Datatype for argument datatype is a null datatype
3: srun: error: nid00217: tasks 1-3: Aborted
3: srun: launch/slurm: _step_signal: Terminating StepId=56572294.33
3: slurmstepd: error: *** STEP 56572294.33 ON nid00217 CANCELLED AT 2022-03-22T10:03:11 ***
3: srun: error: nid00217: tasks 0,4: Terminated
3: srun: Force Terminated StepId=56572294.33
1/1 Test #3: test_rearr .......................***Failed 5.43 sec
```

I only see occurrences of `PMPI_Type_size` in the darshan source code; I built darshan with support to autoperf with `--enable-apmpi-mod`.
Let me know if you want more details from the user, I can probably put you in contact if needed.

Edit: last week we did a maintenance on Cori, after which darshan was upgraded from 3.2.1 to 3.3.1, and the user seems to have been experiencing this only recently, so maybe the issue is in a recent commit after 3.2.1?

**Comment by jedwards4b** [comment]
I am the user and am happy to help if I can.

**Comment by carns** [comment]
Thanks for the bug report @bebosudo and @jedwards4b !

The description and suggested fix make perfect sense.  We'll create a small reproducer in our test environment and get it addressed.  Another developer on our team recently encountered the same problem in a different code base but it didn't occur to us that it would impact Darshan also.

We are going to confirm on our end too, but do you happen to know what MPI build/version exhibits this problem?  We can guard against it regardless of what MPI is being used, of course, but it might help with testing to have a definitive build that trips on this case.

**Comment by bebosudo** [comment]
Currently the default mpi on Cori is `cray-mpich/7.7.19`, but I don't know if that's what @jedwards4b is using.

**Comment by jedwards4b** [comment]
Yes it was.

**Comment by dqwu** [comment]
There is a similar issue in SCORPIO that MPI_Alltoallw() called with MPI_DATATYPE_NULL might fail
with some MPI libs. A workaround is to pass MPI_CHAR instead of MPI_DATATYPE_NULL to MPI_Alltoallw()
for all MPI libs. See https://github.com/E3SM-Project/scorpio/pull/464

Some related threads:
https://users.open-mpi.narkive.com/U5gWUuXD/ompi-mpi-datatype-null-and-mpi-alltoallw
https://lists.mcs.anl.gov/pipermail/mpich-discuss/2009-April/004861.html

**Comment by jedwards4b** [comment]
@dqwu I understand that there are methods to work around the issue.   But it's a workaround and not a fix since MPI_DATATYPE_NULL is an acceptable argument to MPI_Alltoallw.

**Comment by carns** [comment]
@jedwards4b can you share the exact call that is being made by the application when the Darshan problem is triggered?

I was trying to make a simple reproducer with `MPI_File_write()`, but I can't trigger the problem.  In the examples I am trying, the underlying `MPI_File_write()` is returning an error with MPI_DATATYPE_NULL, which means the portion of the Darshan wrapper (see https://github.com/darshan-hpc/darshan/blob/main/darshan-runtime/lib/darshan-mpiio.c#L309) that would inspect the datatype is skipped.

I'm just trying something like this right now, which returns an error code (with or without Darshan): 
```
ret = MPI_File_write(fh, buffer, 0, MPI_DATATYPE_NULL, &status);
```
There must be at least one MPI_File* call that will succeed with the datatype set to MPI_DATATYPE_NULL to be able to trigger this, but I'm not sure what it is.

**Comment by jedwards4b** [comment]
I believe that the call is to MPI_Alltoallv,  I assume that Darshan includes a call to MPI_Type_size in the PMPI there.

**Comment by carns** [comment]
Ah right, I forgot that the AutoPerf APMPI module (https://github.com/argonne-lcf/autoperf/blob/main/apmpi/lib/darshan-apmpi.c) is active on Cori.  It makes `MPI_Type_size` calls that look like they might be easier to trigger than the ones in Darshan's base MPIIO module (https://github.com/darshan-hpc/darshan/blob/main/darshan-runtime/lib/darshan-mpiio.c).

I'm still struggling to reproduce the problem on Cori though; we can add guards in Darshan and AutoPerf, but I would like to be able to test the code path.  I've tried a few variations, with my latest attempt looking like this:

```
#include <stdio.h>
#include <mpi.h>

int main(int argc, char* argv[])
{
    int ret;
    int sendcounts[1] = {0};
    int sdispls[1]    = {0};
    int recvcounts[1] = {0};
    int rdispls[1]    = {0};
    char rbuffer[1]    = {0};

    MPI_Init(&argc, &argv);

    ret = MPI_Alltoallv(MPI_IN_PLACE, sendcounts, sdispls, MPI_DATATYPE_NULL, rbuffer,
                        recvcounts, rdispls, MPI_DATATYPE_NULL, MPI_COMM_WORLD);
    if(ret != MPI_SUCCESS)
        fprintf(stderr, "Error: MPI_Alltoallv(); continuing...\n");

    MPI_Finalize();
    return 0;
}
```
AutoPerf ought to call `MPI_Type_size()` in it's wrapper after the true `MPI_Alltoallv()` function completes, but the problem is that the latter just aborts, so it never gets to the `MPI_Type_size()` call in AutoPerf (this happens with or without Darshan linked in):

```
Rank 0 [Fri Apr  1 03:30:20 2022] [c3-0c1s10n0] Fatal error in PMPI_Alltoallv: Invalid datatype, error stack:
PMPI_Alltoallv(683): MPI_Alltoallv(sbuf=MPI_IN_PLACE, scnts=0x7fffffff70a0, sdispls=0x7fffffff70c0, MPI_DATATYPE_NULL, rbuf=0x7fffffff7120, rcnts=0x7fffffff70e0, rdispls=0x7fffffff7100, MPI_DATATYPE_NULL, MPI_COMM_WORLD) failed
PMPI_Alltoallv(630): Datatype for argument recvtype is a null datatype
srun: error: nid00680: task 0: Aborted
srun: launch/slurm: _step_signal: Terminating StepId=56876481.0
```
I'm just running that with one process, though that looks like an error that would happen regardless of the number of ranks involved.

@jedwards4b any idea what else might help trigger it?  If we can isolate something I'll put a reproducer in our regression tests for posterity.

**Comment by carns** [comment]
Maybe if I set `MPI_Errhandler_set(MPI_COMM_WORLD,MPI_ERRORS_RETURN);` it would proceed past the underlying PMPI_Alltoallv() to trip the MPI_Type_size() in the wrapper?

**Comment by carns** [comment]
Well, to follow up my own question, if I do that then the MPI_Alltoallv() does fail gracefully, but then so does the MPI_Type_size() apparently, so I'm still not really triggering the problem.

**Comment by jedwards4b** [comment]
git clone https://github.com/NCAR/ParallelIO.git
mkdir ParallelIO/bld
cd ParallelIO/bld

 \# Your cmake options may differ, this is for cori-haswell at nersc
CC=cc cmake -DCMAKE_SYSTEM_NAME=Catamount -DCMAKE_BUILD_TYPE=DEBUG -D PnetCDF_PATH:STRING="/opt/cray/pe/parallel-netcdf/1.12.2.1/INTEL/19.1"  -D NetCDF_PATH:PATH=/opt/cray/pe/netcdf-hdf5parallel/4.8.1.1/INTEL/19.1 -DPIO_ENABLE_TIMING=OFF ../

make tests
ctest -VV -I 3,3
```
UpdateCTestConfiguration  from :/global/homes/j/jedwards/pio2_0/bld/DartConfiguration.tcl
Parse Config file:/global/homes/j/jedwards/pio2_0/bld/DartConfiguration.tcl
UpdateCTestConfiguration  from :/global/homes/j/jedwards/pio2_0/bld/DartConfiguration.tcl
Parse Config file:/global/homes/j/jedwards/pio2_0/bld/DartConfiguration.tcl
Test project /global/homes/j/jedwards/pio2_0/bld
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 3
    Start 3: test_rearr

3: Test command: /global/homes/j/jedwards/pio2_0/cmake/mpiexec.nersc "5" "/global/homes/j/jedwards/pio2_0/bld/tests/cunit/test_rearr"
3: Test timeout computed to be: 600
3: Rank 1 [Fri Apr  1 08:12:57 2022] [c11-0c2s8n0] Fatal error in PMPI_Type_size: Invalid datatype, error stack:
3: PMPI_Type_size(119): MPI_Type_size(MPI_DATATYPE_NULL) failed
3: PMPI_Type_size(69).: Datatype for argument datatype is a null datatype
3: Rank 2 [Fri Apr  1 08:12:57 2022] [c11-0c2s8n0] Fatal error in PMPI_Type_size: Invalid datatype, error stack:
3: PMPI_Type_size(119): MPI_Type_size(MPI_DATATYPE_NULL) failed
3: PMPI_Type_size(69).: Datatype for argument datatype is a null datatype
3: Rank 3 [Fri Apr  1 08:12:57 2022] [c11-0c2s8n0] Fatal error in PMPI_Type_size: Invalid datatype, error stack:
3: PMPI_Type_size(119): MPI_Type_size(MPI_DATATYPE_NULL) failed
3: PMPI_Type_size(69).: Datatype for argument datatype is a null datatype
3: srun: error: nid02272: tasks 1-3: Aborted
3: srun: launch/slurm: _step_signal: Terminating StepId=56884555.4
3: slurmstepd: error: *** STEP 56884555.4 ON nid02272 CANCELLED AT 2022-04-01T08:12:57 ***
3: srun: error: nid02272: tasks 0,4: Terminated
3: srun: Force Terminated StepId=56884555.4
1/1 Test #3: test_rearr .......................***Failed    1.86 sec

0% tests passed, 1 tests failed out of 1
```

**Comment by jedwards4b** [comment]
Here is a little more info:

**Comment by carns** [comment]
Perfect!  Thanks @jedwards4b 

The ctest command didn't work for me (probably requires some other environment step) but the test is easy to run in a standalone slurrm script after compiling:

```
#!/bin/bash -l
#SBATCH -N 1         #Use 1 nodes
#SBATCH -t 00:10:00  #Set 10 minute time limit
#SBATCH -p debug   #Submit to the debug 'partition'
#SBATCH -L SCRATCH   #Job requires $SCRATCH file system
#SBATCH -C haswell   #Use Haswell nodes
#SBATCH --mail-user=carns@mcs.anl.gov
#SBATCH --mail-type=ALL
 
ulimit -c unlimited
srun -n 5 ParallelIO/bld/tests/cunit/test_rearr
```
The resulting core file shows this:
```
#0  0x0000155552540360 in raise () from /lib64/libc.so.6
#1  0x0000155552541941 in abort () from /lib64/libc.so.6
#2  0x00001555538b9868 in MPID_Abort ()
   from /opt/cray/pe/lib64/libmpich_intel.so.3
#3  0x0000155553841812 in MPIR_Handle_fatal_error ()
   from /opt/cray/pe/lib64/libmpich_intel.so.3
#4  0x000015555384194e in MPIR_Err_return_comm ()
   from /opt/cray/pe/lib64/libmpich_intel.so.3
#5  0x00001555537ac45d in PMPI_Type_size ()
   from /opt/cray/pe/lib64/libmpich_intel.so.3
#6  0x0000155554aa1a84 in PMPI_Alltoallw ()
   from /global/common/cori_cle7up03/software/darshan/3.3.1/lib/libdarshan.so
#7  0x000000002003047b in pio_swapm (sendbuf=0x2, sendcounts=0x7fffffff4a10, 
    sdispls=0x0, sendtypes=0x155552540360 <raise+272>, recvbuf=0x0, 
    recvcounts=0x7fffffff4a10, rdispls=0x7fffffff66a0, 
    recvtypes=0x7fffffff6660, comm=-2080374780, fc=0x1000008da30)
    at /global/homes/p/pcarns/working/dbg/dbg-darshan-mpi_datatype_null-2022-03/ParallelIO/src/clib/pio_spmd.c:128
#8  0x00000000200322f9 in rearrange_comp2io (ios=0x2, iodesc=0x7fffffff4a10, 
    sbuf=0x0, rbuf=0x155552540360 <raise+272>, nvars=0)
    at /global/homes/p/pcarns/working/dbg/dbg-darshan-mpi_datatype_null-2022-03/ParallelIO/src/clib/pio_rearrange.c:982
#9  0x000000002000c3cc in test_rearrange_comp2io (test_comm=-2080374780, 
    my_rank=2)
    at /global/homes/p/pcarns/working/dbg/dbg-darshan-mpi_datatype_null-2022-03/ParallelIO/tests/cunit/test_rearr.c:1047
#10 0x000000002000d821 in run_no_iosys_tests (my_rank=2, test_comm=-2080374780)
    at /global/homes/p/pcarns/working/dbg/dbg-darshan-mpi_datatype_null-2022-03/ParallelIO/tests/cunit/test_rearr.c:1291
#11 0x000000002000e95b in main (argc=1, argv=0x7fffffff6c98)
    at /global/homes/p/pcarns/working/dbg/dbg-darshan-mpi_datatype_null-2022-03/ParallelIO/tests/cunit/test_rearr.c:1560
```

That's definitely tripping on the MPI_Type_size() call in an APMPI wrapper.  I don't know exactly what the arguments are, but one wrinkle is that it's actually in MPI_Alltoallw() instead of MPI_Alltoallv() (possibly easier to trigger there because there is a potential array of types instead of just one).

I'll probably transfer this ticket to the AutoPerf repo; we'll want to fix there first, but once the fix is confirmed we can do the same thing to other occurrences in the core darshan modules too.

**Comment by carns** [comment]
Here is a bare minimum reproducer that can trigger the problem with a single process on Cori with the default environment.  It also triggers it on my laptop with MPICH 3.4.1 and Darshan origin/main with --enable-apmpi-mod configured:

```
#include <stdio.h>
#include <mpi.h>

int main(int argc, char* argv[])
{
    int          ret;
    int          sendcounts[1] = {0};
    int          sdispls[1]    = {0};
    int          recvcounts[1] = {0};
    int          rdispls[1]    = {0};
    char         rbuffer[1]    = {0};
    char         sbuffer[1]    = {0};
    MPI_Datatype sendtypes[1]  = {MPI_DATATYPE_NULL};
    MPI_Datatype recvtypes[1]  = {MPI_DATATYPE_NULL};
    int          nprocs        = 0;

    MPI_Init(&argc, &argv);

    MPI_Comm_size(MPI_COMM_WORLD, &nprocs);

    if (nprocs != 1) {
        fprintf(stderr,
                "Error: this test program must be executed with exactly one "
                "process.\n");
        return (-1);
    }

#if 0
    MPI_Errhandler_set(MPI_COMM_WORLD,MPI_ERRORS_RETURN);
#endif

    ret = MPI_Alltoallw(sbuffer, sendcounts, sdispls, sendtypes, rbuffer,
                        recvcounts, rdispls, recvtypes, MPI_COMM_WORLD);

    if (ret != MPI_SUCCESS)
        fprintf(stderr, "Error: MPI_Alltoallw(); continuing...\n");

    MPI_Finalize();
    return 0;
}
```

**Comment by carns** [comment]
Cori error:
```
Rank 0 [Fri Apr  1 14:35:53 2022] [c3-0c2s12n3] Fatal error in PMPI_Type_size: Invalid datatype, error stack:
PMPI_Type_size(119): MPI_Type_size(MPI_DATATYPE_NULL) failed
PMPI_Type_size(69).: Datatype for argument datatype is a null datatype
srun: error: nid00755: task 0: Aborted (core dumped)
srun: launch/slurm: _step_signal: Terminating StepId=56895150.0
```
Laptop error:
```
Fatal error in PMPI_Type_size: Invalid datatype, error stack:
PMPI_Type_size(113): MPI_Type_size(MPI_DATATYPE_NULL) failed
PMPI_Type_size(63).: Datatype for argument datatype is a null datatype
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=134883075
:
system msg for write_line failure : Bad file descriptor
```

**Comment by shanedsnyder** [comment]
This should be fixed now, and will be included in Darshan 3.4.0 (which should be out in the next day or so).
